source/utils/paragraphHelper.py

```diff
@@ -0,0 +1,271 @@
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2022 NV Access Limited, Rob Meredith
+# This file may be used under the terms of the GNU General Public License, version 2 or later.
+# For more details see: https://www.gnu.org/licenses/gpl-2.0.html
+
+import api
+import speech
+import controlTypes
+import textInfos
+import tones
+import config
+from NVDAObjects.window.winword import WordDocumentTextInfo
+from NVDAObjects.window.winword import BrowseModeWordDocumentTextInfo
+from NVDAObjects.UIA import UIATextInfo
+from displayModel import EditableTextDisplayModelTextInfo
+from typing import (
+	Tuple,
+	List,
+)
+from enum import IntEnum
+
+MAX_LINES = 250  # give up after searching this many lines
+
+
+class _ErrorTone(IntEnum):
+	FREQUENCY_HZ = 1000
+	DURATION_MS = 30
+
+
+class _Offset(IntEnum):
+	PREVIOUS_LINE = -1
+	NEXT_LINE = 1
+
+
+def nextParagraphStyle() -> config.featureFlag.FeatureFlag:
+	from config.featureFlagEnums import ParagraphNavigationFlag
+	flag: config.featureFlag.FeatureFlag = config.conf["documentNavigation"]["paragraphStyle"]
+	numStyles = len(ParagraphNavigationFlag.__members__)
+	newEnumVal = flag.calculated().value + 1
+	if newEnumVal > numStyles:
+		newEnumVal = ParagraphNavigationFlag.DEFAULT.value + 1  # wrap around, skip DEFAULT
+
+	return ParagraphNavigationFlag(newEnumVal)
+
+
+def getTextInfoAtCaret() -> textInfos.TextInfo:
+	# returns None if not editable text or document
+	ti = None
+	focus = api.getFocusObject()
+	if (focus.role == controlTypes.Role.EDITABLETEXT) or (focus.role == controlTypes.Role.DOCUMENT):
+		try:
+			ti = focus.makeTextInfo(textInfos.POSITION_CARET)
+		except (NotImplementedError, RuntimeError):
+			pass
+	return ti
+
+
+def _isAcceptableTextInfo(ti: textInfos.TextInfo) -> bool:
+	acceptable = True
+	# disallow if in a Word document and not using UIA, as Word has performance issues
+	if isinstance(ti, WordDocumentTextInfo) or isinstance(ti, BrowseModeWordDocumentTextInfo):
+		acceptable = False
+	# disallow if EditableTextDisplayModelTextInfo, as has performance issues (TextPad for example)
+	if isinstance(ti, EditableTextDisplayModelTextInfo):
+		acceptable = False
+	return acceptable
+
+
+def _isLastLineOfParagraph(line: str) -> bool:
+	stripped = line.strip(' \t')
+	return stripped.endswith('\r') or stripped.endswith('\n')
+
+
+def _splitParagraphIntoChunks(paragraph: str) -> List[str]:
+	CHUNK_SIZE = 2048
+	SENTENCE_TERMINATOR = ". "
```

Comment:
Sentences can end with other punctuations: question mark (?), exclaim (!)...

And also sentence end also depends on the language, e.g. arabic dot and ideographic dot used instead of western latin dot. Also I think that in east-asian texts, there is no space joined to the ideographic dot.

@cary-rowen or other people from east-Asia or Arabic community:
Please confirm or not my writings. Thanks.

Reply:
@CyrilleB79 Yes, I considered sentence termination in other languages briefly, and I'm no expert on this. But, the chunk splitter is only a helper, so even if it fails, everything will still work. It is just a bit more efficient if smaller chunks are produced. This is why I didn't worry about question mark, exclamation point, etc.

Reply:
Would it be possible to split after every N spaces instead?

Reply:
@seanbudd Breaking on spaces would probably cause odd pauses in speech.
 I have added support for a frozen set of sentence terminators, which currently includes period, exclamation point, and question mark. The rule is that a space must follow any terminator. Can I push this, or do I need to reset this pr as draft?

Reply:
Feel free to push this.
I think it would be good for if we kept a set of speech terminators in a common location.
To be able to reliably do this I imagine we need `nltk` or something similar